File: cpp/0654最大二叉树.cpp
#include<vector>
#include"mybitree.h"

// 左闭右开
TreeNode* Traversal(vector<int> &nums,int left,int right)
{
    if(left>=right)
        return nullptr;

    int val = nums[left];
    int idx = left;
    for (int i = left + 1; i < right; ++i)
    {
        if(val<nums[i])
        {
            idx = i;
            val = nums[i];
        }
    }

    TreeNode *root = new TreeNode(val);

    root->left = Traversal(nums, left, idx);
    root->right = Traversal(nums, idx+1, right);
    return root;
}

TreeNode *constructMaximumBinaryTree(vector<int> &nums)
{
    return Traversal(nums, 0, nums.size());
}

File: python/main/0654最大二叉树.py
from binarytree import TreeNode


def constructMaximumBinaryTree(nums:list) -> TreeNode:
    if(not nums):
        return None

    val=max(nums)
    idx=nums.index(val)
    root=TreeNode(val)

    root.left=constructMaximumBinaryTree(nums[:idx])
    root.right=constructMaximumBinaryTree(nums[idx+1:])

    return root
